### PR TITLE
OCPBUGS-34399: Exposing chunksize variable to utilize docker registry config

### PR DESCRIPTION
**Pull request by deepsm007**
Trying to work around a bug for r2 integration with s3 bucket, exposing chunksize variable by utilizing the docker registry S3 driver configuration option: https://docker-docs.uclv.cu/registry/storage-drivers/s3/ 

/hold for https://github.com/openshift/api/pull/1864 to merge

https://issues.redhat.com//browse/IR-471 

/cc @flavianmissi

### CODE REVIEW

File: pkg/operator/starter.go [hunk 1]
```diff
@@ -66,8 +68,21 @@ func RunOperator(ctx context.Context, kubeconfig *restclient.Config) error {
 	if err != nil {
 		klog.Warningf("unable to get owner reference (falling back to namespace): %v", err)
 	}
+
 	eventRecorder := events.NewKubeRecorder(kubeClient.CoreV1().Events(defaults.ImageRegistryOperatorNamespace), "image-registry-operator", controllerRef)
 
+	desiredVersion := status.VersionForOperatorFromEnv()
```
Comment (flavianmissi) [inline pkg/operator/starter.go:74]: Following the implementation of `VersionForOperatorFromEnv` I see that the value comes from the env var `OPERATOR_IMAGE_VERSION`, which the [regisitry operator manifest](https://github.com/openshift/cluster-image-registry-operator/blob/master/manifests/07-operator.yaml#L59-L79) does not specify.
Could you please add the `OPERATOR_IMAGE_VERSION` env var to the operator manifest yaml? Here's an example: https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/manifests/09_deployment.yaml#L60-L61. I think that OCP's building tools will replace the value with something other than the placeholder, so you can use the example as is in our manifest.

**Review by flavianmissi — COMMENTED** [review]
Could you separate feature gate support from the chunksize config commit? More specifically I'd like to see all the dependency injection updates as well as feature gate accessor initialization in the same commit. Thanks!

**Comment by flavianmissi** [comment]
Running `make test-unit` locally gives me an error:
```
# k8s.io/apiserver/pkg/endpoints/handlers
vendor/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:204:17: undefined: streaming.NewEncoderWithAllocator
# k8s.io/apiserver/pkg/server/routes
vendor/k8s.io/apiserver/pkg/server/routes/openapi.go:68:89: cannot use oa.Config (variable of type *common.Config) as *common.OpenAPIV3Config value in argument to builder3.BuildOpenAPISpecFromRoutes
# github.com/openshift/cluster-image-registry-operator/pkg/storage/pvc
pkg/storage/pvc/pvc.go:173:15: cannot use corev1.ResourceRequirements{…} (value of type "k8s.io/api/core/v1".ResourceRequirements) as "k8s.io/api/core/v1".VolumeResourceRequirements value in struct literal
```
can you have a look @deepsm007 ?

**Comment by deepsm007** [comment]
https://github.com/openshift/cluster-image-registry-operator/pull/1036 to bump the api, client-go & library-go

**Comment by flavianmissi** [comment]
unit test failing:
```
{  === RUN   TestMakePodTemplateSpecS3CloudFront
    podtemplatespec_test.go:560: expected env var REGISTRY_STORAGE_S3_CHUNKSIZE not found
```
since this env var is optional, should the test really be expecting it?

**Comment by flavianmissi** [comment]
Looks like there might be some permissions missing from the operator as well, seeing this in the logs:
```
W0522 05:55:27.620106       1 reflector.go:539] github.com/openshift/client-go/config/informers/externalversions/factory.go:125: failed to list *v1.FeatureGate: featuregates.config.openshift.io is forbidden: User "system:serviceaccount:openshift-image-registry:cluster-image-registry-operator" cannot list resource "featuregates" in API group "config.openshift.io" at the cluster scope
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-image-registry-operator/1026/pull-ci-openshift-cluster-image-registry-operator-master-e2e-aws-ovn-upgrade/1793105514377252864/artifacts/e2e-aws-ovn-upgrade/gather-extra/artifacts/pods/openshift-image-registry_cluster-image-registry-operator-787b97fc95-pgxmm_cluster-image-registry-operator.log

File: pkg/operator/starter.go [hunk 2]
```diff
@@ -205,6 +228,7 @@ func RunOperator(ctx context.Context, kubeconfig *restclient.Config) error {
 	go azureStackCloudController.Run(ctx)
 	go azurePathFixController.Run(ctx.Done())
 	go metricsController.Run(ctx)
+	go featureGateAccessor.Run(ctx)
```
Comment (flavianmissi) [inline pkg/operator/starter.go:231]: This is already called on line 81 - did we really mean to call it again here?

File: pkg/storage/util/util.go [hunk 3]
```diff
@@ -132,3 +136,13 @@ func GenerateStorageName(listers *regopclient.StorageListers, additionalInfo ...
 
 	return strings.ToLower(name), nil
 }
+
+// Helper function to test featuregate ChunkSizeMiB
+func ChunkSizeMiBFeatureGateEnabled(featureGateAccessor featuregates.FeatureGateAccess) (bool, error) {
```
Comment (flavianmissi) [inline pkg/storage/util/util.go:141]: Where is this function used? 🤔

**Review by flavianmissi — COMMENTED** [review]
Where is the change that actually adds the chunksize env var to image registry pod spec? I think maybe that's why that unit test was failing? In that case, maybe we should add it back (sorry...)?

File: pkg/storage/util/util.go (hunk 3 above)
Comment (deepsm007) [reply pkg/storage/util/util.go:141]: here --> https://github.com/openshift/cluster-image-registry-operator/pull/1026/commits/933c26e3290824dc55319dd270ff07ad3ab2c046#diff-fcbc7c17a683424f7d245a7af42669bc401010621d8fccfa1da7860e152e20d3R512

File: pkg/operator/starter.go [hunk 4]
```diff
@@ -68,6 +72,24 @@ func RunOperator(ctx context.Context, kubeconfig *restclient.Config) error {
 	}
 	eventRecorder := events.NewKubeRecorder(kubeClient.CoreV1().Events(defaults.ImageRegistryOperatorNamespace), "image-registry-operator", controllerRef)
 
+	desiredVersion := status.VersionForOperatorFromEnv()
+	missingVersion := "0.0.1-snapshot"
+
+	// By default, this will exit(0) the process if the featuregates ever change to a different set of values.
+	featureGateAccessor := featuregates.NewFeatureGateAccess(desiredVersion, missingVersion, configInformers.Config().V1().ClusterVersions(), configInformers.Config().V1().FeatureGates(), eventRecorder)
+
+	go featureGateAccessor.Run(ctx)
```
Comment (flavianmissi) [inline pkg/operator/starter.go:81]: should we give `ctx.Done` to `Run` like the other calls to `Run` do down below? i.e:
```
go featureGateAccessor.Run(ctx.Done())
```

Comment (flavianmissi) [reply pkg/operator/starter.go:81]: Never mind - `ctx.Done()` is a channel, and `fetureGateAccessor.Run` takes a regular context parameter.

File: pkg/operator/starter.go [hunk 5]
```diff
@@ -190,7 +214,6 @@ func RunOperator(ctx context.Context, kubeconfig *restclient.Config) error {
 	kubeInformersForOpenShiftConfig.Start(ctx.Done())
 	kubeInformersForOpenShiftConfigManaged.Start(ctx.Done())
 	kubeInformersForKubeSystem.Start(ctx.Done())
-	configInformers.Start(ctx.Done())
```
Comment (flavianmissi) [inline pkg/operator/starter.go:193]: let's try adding this line back here? I know it's already called higher up on this file, but let's just test and see if changes anything.

Comment (deepsm007) [reply pkg/operator/starter.go:193]: add it multiple times or remove the one from above and add here?

Comment (flavianmissi) [reply pkg/operator/starter.go:193]: Add it multiple times

**Comment by flavianmissi** [comment]
vsphere failures indicate that the feature gate detection code is interfering with graceful shutdown:
```
E0513 19:59:57.528191       1 starter.go:89] timed out waiting for FeatureGate detection
W0513 19:59:57.528346       1 builder.go:130] graceful termination failed, controllers failed with error: timed out waiting for FeatureGate detection
```
link to logs: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-image-registry-operator/1026/pull-ci-openshift-cluster-image-registry-operator-master-e2e-vsphere/1790093611942023168/artifacts/e2e-vsphere/gather-extra/artifacts/pods/openshift-image-registry_cluster-image-registry-operator-65f88ff885-mk88n_cluster-image-registry-operator.log
can you look into it @deepsm007 ? the operator needs to be able to shutdown gracefully as it previously did.

**Comment by deepsm007** [comment]
> vsphere failures indicate that the feature gate detection code is interfering with graceful shutdown:
> 
> ```
> E0513 19:59:57.528191       1 starter.go:89] timed out waiting for FeatureGate detection
> W0513 19:59:57.528346       1 builder.go:130] graceful termination failed, controllers failed with error: timed out waiting for FeatureGate detection
> ```
> 
> link to logs: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-image-registry-operator/1026/pull-ci-openshift-cluster-image-registry-operator-master-e2e-vsphere/1790093611942023168/artifacts/e2e-vsphere/gather-extra/artifacts/pods/openshift-image-registry_cluster-image-registry-operator-65f88ff885-mk88n_cluster-image-registry-operator.log can you look into it @deepsm007 ? the operator needs to be able to shutdown gracefully as it previously did.

That run is from old commit, out of the last 4 runs 2 passed and 2 failed (not related to the PR).
[This](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-image-registry-operator/1026/pull-ci-openshift-cluster-image-registry-operator-master-e2e-vsphere/1795109388092968960/artifacts/e2e-vsphere/gather-extra/artifacts/pods/openshift-image-registry_cluster-image-registry-operator-658f979877-cs5tc_cluster-image-registry-operator.log) is the most recent run.

**Comment by flavianmissi** [comment]
hypershift failures are unrelated
/override ci/prow/e2e-hypershift
/override ci/prow/e2e-hypershift-conformance

**Pull request merged**